src/core/server/plugins/plugin.ts

```diff
@@ -88,11 +91,16 @@ export class PluginWrapper<
    * is the contract returned by the dependency's `setup` function.
    */
   public setup(
-    setupContext: CoreSetup<TPluginsStart>,
+    setupContext: CoreSetup<TPluginsStart> | CorePreboot,
     plugins: TPluginsSetup
   ): TSetup | Promise<TSetup> {
     this.instance = this.createPluginInstance();
-    return this.instance.setup(setupContext, plugins);
+
+    if (this.isPrebootPluginInstance(this.instance)) {
+      return this.instance.setup(setupContext as CorePreboot, plugins);
+    }
+
+    return this.instance.setup(setupContext as CoreSetup, plugins);
```

Comment:
__note:__ my Typescript-fu wasn't strong enough to find a better typed option to cover this use case, happy to apply any other suggestions here!

Reply:
> return IDK;

Yeah, this particular thing has stopped me from writing this guard just for the sake of writing it. Cause if we don't validate all properties of the contracts it won't be much better than simple `as CoreSetup/CorePreboot`, but validating all properties isn't worth it, it seems.

Reply:
@pgayvallet I was experimenting with splitting `PluginWrapper` into preboot and standard "entities" as per your request, and we can do something similar to this (semi-pseudo code, but I have a WIP that seems to work well too). Does it look like something you had in mind?

If so, then the only thing that concerns me is that it requires quite a bit of changes (in tests as well) and I'd probably be more comfortable with making them in a follow up. 

If it's not what you meant, then I'm all ears :slightly_smiling_face: 

```ts
export class PluginWrapper {
  public readonly path: string;
  ...... other manifest/general properties and methods .....

  // The instance type (`OpaquePlugin`) is now not bound to type, it doesn't have to. 
  // interface OpaquePlugin { setup: Function }
  private instance?: OpaquePlugin;

  public getInstance<T extends PluginType>(): PluginInstance<T> {
    if (this.instance === undefined) {
      this.instance = this.createPluginInstance();
    }

    return this.instance as PluginInstance<T>;
  }

  private createPluginInstance() {
    ......
    // At this point we don't care about proper typing of the initializer, we cannot use
    // types for anything useful here:
    // type OpaquePluginInitializer = (core: PluginInitializerContext) => OpaquePlugin;
    const { plugin: initializer } = pluginDefinition as { plugin: OpaquePluginInitializer };
    ......
    const instance = initializer(this.initializerContext);
    ......
    // Base on the type declared in manifest we create a proper instance wrapper (see below).
    return this.manifest.type === PluginType.preboot
      ? new PrebootPluginInstance(this.instance as PrebootPlugin)
      : new StandardPluginInstance(this.name, this.instance as Plugin);
  }
}
-------------
export class PrebootPluginInstance<TSetup = unknown, TPluginsSetup extends object = object> {
  constructor(private readonly innerInstance: PrebootPlugin<TSetup, TPluginsSetup>) {}
  public setup(setupContext: CorePreboot, plugins: TPluginsSetup): TSetup | Promise<TSetup>;
  public async stop();
}
-------------
export class StandardPluginInstance<
  TSetup = unknown,
  TStart = unknown,
  TPluginsSetup extends object = object,
  TPluginsStart extends object = object
> {
  public readonly startDependencies = this.startDependencies$.pipe(first()).toPromise();
  constructor(
    private readonly innerInstance:
      | Plugin<TSetup, TStart, TPluginsSetup, TPluginsStart>
      | AsyncPlugin<TSetup, TStart, TPluginsSetup, TPluginsStart>
  ) {}
  public setup(setupContext: CoreSetup, plugins: TPluginsSetup): TSetup | Promise<TSetup>;
  public start(startContext: CoreStart, plugins: TPluginsStart): TStart | Promise<TStart>;
  public async stop();
}
```

And in the `PluginSystem` we can do something like this:
```ts
  public async setupPlugins(
    deps: T extends PluginType.preboot ? PluginsServicePrebootSetupDeps : PluginsServiceSetupDeps
  ): Promise<Map<string, unknown>> {
   .......
   if (this.type === PluginType.preboot) {
      const setupContext = createPluginPrebootSetupContext(....);
      plugin
        .getInstance<PluginType.preboot>()
        .setup(setupContext, pluginDepContracts);
    } else {
      const setupContext = createPluginSetupContext(....);
      plugin
        .getInstance<PluginType.standard>()
        .setup(setupContext, pluginDepContracts);
    }
   ///// OR `getInstance` to `asSomething` method that can do run-time validation too.
   if (this.type === PluginType.preboot) {
      const setupContext = createPluginPrebootSetupContext(....);
      plugin.asPreboot().setup(setupContext, pluginDepContracts);
    } else {
      const setupContext = createPluginSetupContext(....);
      plugin.asStandard().setup(setupContext, pluginDepContracts);
    }
  }
```

Reply:
> And in the PluginSystem we can do something like this:

yeah, something like that. As for implementation, I'd either create two separate methods `setupPlugins/setupPrebootPlugins` or refactor `PluginSystem` to have separate instances for `preboot/standard` plugins.

Reply:
> refactor PluginSystem to have separate instances for preboot/standard plugins.

I see if I can find a good way to do that. The main difficulty is different dependencies and different `createPluginPrebootSetupContext` used to setup these plugins.

Reply:
Okay I gave a quick try to this idea and would like to propose to postpone it for the follow-up, it's quite a large change unfortunately and PR is already `+5,752 −1,217`, I'm afraid it'll be even harder to review it with these improvements.

The rest of the feedback should be addressed at this point.

How does that sound?

Reply:
Will be handling this in the scope of https://github.com/elastic/kibana/pull/106218